torch/_inductor/autotuner/model.py

```diff
@@ -0,0 +1,801 @@
+        res.append(dep_feature)
+    for _ in range(rw_lim - len(res)):
+        res.append(make_dep_feature(ndims_lim=ndims_lim))
+    return res
+
+
+### model arch related
+
+
+class ModelType(enumerate):
+    XGB_BASELINE = 0
+    NN_POINTWISE = 1
+    NN_PAIRWISE = 2
+    NN_PAIRWISE_SMALL = 3
+
+
+class Autotuner_FFN(nn.Module):
+    def __init__(self, model_cfg):
+        super().__init__()
+
+        self.kernel_category_embedding = torch.nn.Embedding(
+            num_embeddings=model_cfg["kernel_category_cnt"],
+            embedding_dim=model_cfg["kernel_category_embed_dim"],
+        )
+        self.num_of_loops_embedding = torch.nn.Embedding(
+            num_embeddings=model_cfg["num_of_loops_cnt"],
+            embedding_dim=model_cfg["num_of_loops_embed_dim"],
+        )
+
+        self.hidden_dim = [model_cfg["feature_dim"]] + model_cfg["hidden_dim"] + [1]
+        self.num_layers = len(self.hidden_dim) - 1
+
+        self.op_bag_ln = nn.ModuleList(
+            [
+                nn.Linear(1, model_cfg["op_embed_dim"])
+                for i in range(model_cfg["op_cnt"])
+            ]
+        )
+        self.is_contiguous_ln = torch.nn.Embedding(
+            num_embeddings=2, embedding_dim=model_cfg["bool_embed_dim"]
+        )
+        self.is_scalar_ln = torch.nn.Embedding(
+            num_embeddings=2, embedding_dim=model_cfg["bool_embed_dim"]
+        )
+        self.is_indirect_ln = torch.nn.Embedding(
+            num_embeddings=2, embedding_dim=model_cfg["bool_embed_dim"]
+        )
+
+        self.layers = nn.ModuleList(
+            [
+                nn.Linear(self.hidden_dim[i], self.hidden_dim[i + 1])
+                for i in range(self.num_layers)
+            ]
+        )
+        self.use_norm = model_cfg["use_norm"]
+        self.norms = nn.ModuleList(
+            [nn.LayerNorm(self.hidden_dim[i + 1]) for i in range(self.num_layers - 1)]
+        )
+
+        torch.nn.init.xavier_normal_(self.kernel_category_embedding.weight)
+        torch.nn.init.xavier_normal_(self.num_of_loops_embedding.weight)
+        torch.nn.init.xavier_normal_(self.is_contiguous_ln.weight)
+        torch.nn.init.xavier_normal_(self.is_scalar_ln.weight)
+        torch.nn.init.xavier_normal_(self.is_indirect_ln.weight)
+        for layer in list(self.op_bag_ln) + list(self.layers):
+            torch.nn.init.xavier_normal_(layer.weight)
+            torch.nn.init.zeros_(layer.bias)
+
+        self.activation = model_cfg["activation"]
+
+    def forward(self, x):
+        if not isinstance(x, list):
+            x = [x]
+        x_kernel_category = np.array([i.kernel_feature.kernel_category for i in x])
+        x_kernel_category = torch.from_numpy(x_kernel_category).to("cuda").long()
+        x_kernel_category = self.kernel_category_embedding(x_kernel_category)
+
+        x_num_of_loops = np.array([i.kernel_feature.num_of_loops for i in x])
+        x_num_of_loops = torch.from_numpy(x_num_of_loops).to("cuda").long()
```

Comment:
Throughout the file we are hardcoding cuda, we should make this work with other devices (for example of cuda isn't available).